File: stdlib/lua/hull/web/tests/test_flash.lua
-- test_flash.lua - Tests for hull.web.flash
--
-- Covers input validation and the HX-Trigger path. The session-backed
-- paths (set / consume) require a live SQLite session store; they are
-- exercised end-to-end in examples/hypermedia_todo/tests/test_app.lua.

local flash = require('hull.web.flash')
local json  = require('hull.json')

local pass = 0
local fail = 0

local function test(name, fn)
    local ok, err = pcall(fn)
    if ok then
        pass = pass + 1
    else
        fail = fail + 1
        print("FAIL: " .. name .. ": " .. tostring(err))
    end
end

local function assert_eq(a, b, msg)
    if a ~= b then
        error((msg or "") .. " expected " .. tostring(b) .. ", got " .. tostring(a))
    end
end

-- Fake res object that captures headers in a table.
local function fake_res()
    local headers = {}
    return {
        header = function(self, k, v) headers[k] = v end,
        _headers = headers,
    }
end

-- ── trigger ──────────────────────────────────────────────────────────

test("trigger sets HX-Trigger header with default kind=info", function()
    local res = fake_res()
    flash.trigger(res, "Saved.")
    local val = res._headers["HX-Trigger"]
    assert(val, "header set")
    local payload = json.decode(val)
    assert_eq(payload.flash.text, "Saved.")
    assert_eq(payload.flash.kind, "info")
end)

test("trigger honors custom kind", function()
    local res = fake_res()
    flash.trigger(res, "Bad input", "error")
    local payload = json.decode(res._headers["HX-Trigger"])
    assert_eq(payload.flash.text, "Bad input")
    assert_eq(payload.flash.kind, "error")
end)

test("trigger coerces non-string text", function()
    local res = fake_res()
    flash.trigger(res, 42)
    local payload = json.decode(res._headers["HX-Trigger"])
    assert_eq(payload.flash.text, "42")
end)

test("trigger rejects missing res", function()
    local ok = pcall(flash.trigger, nil, "x")
    assert_eq(ok, false, "should error")
end)

test("trigger rejects missing text", function()
    local ok = pcall(flash.trigger, fake_res(), nil)
    assert_eq(ok, false, "should error")
end)

-- ── set: input validation (no session needed for the error path) ────

test("set rejects missing text", function()
    local req = { ctx = { session_id = "fake", session = {} } }
    local ok, err = pcall(flash.set, req, nil)
    assert_eq(ok, false, "should error")
    assert(tostring(err):find("text is required"),
           "error mentions text required, got: " .. tostring(err))
end)

test("set errors when no session_id in req.ctx", function()
    local req = { ctx = {} }
    local ok, err = pcall(flash.set, req, "hi")
    assert_eq(ok, false, "should error")
    assert(tostring(err):find("no session"),
           "error mentions session, got: " .. tostring(err))
end)

test("set errors when req is nil", function()
    local ok = pcall(flash.set, nil, "hi")
    assert_eq(ok, false, "should error")
end)

-- ── consume: safe-defaults ───────────────────────────────────────────

test("consume returns empty array when no session", function()
    local msgs = flash.consume({ ctx = {} })
    assert_eq(type(msgs), "table")
    assert_eq(#msgs, 0)
end)

test("consume returns empty when req is nil", function()
    local msgs = flash.consume(nil)
    assert_eq(type(msgs), "table")
    assert_eq(#msgs, 0)
end)

-- ── session-backed round trip ───────────────────────────────────────

local session = require('hull.web.middleware.session')
session.init({ ttl = 60 })

local function ctx_for_new_session()
    local sid = session.create({})
    return { ctx = { session_id = sid } }
end

test("set + consume round-trip persists and clears in one cycle", function()
    local req = ctx_for_new_session()
    flash.set(req, "First message.", "info")
    flash.set(req, "Second message.", "success")
    local msgs = flash.consume(req)
    assert_eq(#msgs, 2)
    assert_eq(msgs[1].text, "First message.")
    assert_eq(msgs[1].kind, "info")
    assert_eq(msgs[2].text, "Second message.")
    assert_eq(msgs[2].kind, "success")
    -- Second consume on the same req → empty (one-shot semantics).
    local empty = flash.consume(req)
    assert_eq(#empty, 0)
end)

test("consume across separate requests in the same session", function()
    local sid = session.create({})
    -- Request A: stash.
    flash.set({ ctx = { session_id = sid } }, "across requests")
    -- Request B (separate req table, same session id): consume.
    local msgs = flash.consume({ ctx = { session_id = sid } })
    assert_eq(#msgs, 1)
    assert_eq(msgs[1].text, "across requests")
end)

test("set bounded at FLASH_MAX_ENTRIES (16); excess silently dropped", function()
    local req = ctx_for_new_session()
    for i = 1, 25 do
        flash.set(req, "msg-" .. i, "info")
    end
    local msgs = flash.consume(req)
    assert_eq(#msgs, 16, "queue capped at 16")
    assert_eq(msgs[1].text, "msg-1", "first-in preserved")
    assert_eq(msgs[16].text, "msg-16", "later messages dropped")
end)

test("set truncates text exceeding 4 KiB", function()
    local req = ctx_for_new_session()
    flash.set(req, string.rep("a", 5000), "info")
    local msgs = flash.consume(req)
    assert_eq(#msgs, 1)
    assert_eq(#msgs[1].text, 4096)
end)

return { pass = pass, fail = fail }

File: stdlib/js/hull/web/tests/test_flash.js
// test_flash.js - Tests for hull:web:flash
//
// Covers input validation and the HX-Trigger path. The session-backed
// paths (set / consume) require a live SQLite session store; they are
// exercised end-to-end in examples/hypermedia_todo/tests/test_app.js.

import { flash } from "hull:web:flash";
import { json }  from "hull:json";

let pass = 0;
let fail = 0;

function test(name, fn) {
    try {
        fn();
        pass++;
    } catch (e) {
        fail++;
        console.log(`FAIL: ${name}: ${e.message || e}`);
    }
}

function assertEq(a, b, msg) {
    if (a !== b) {
        throw new Error(`${msg || ""} expected ${b}, got ${a}`);
    }
}

// Fake res object that captures headers in a plain object.
function fakeRes() {
    const headers = {};
    return {
        header(k, v) { headers[k] = v; },
        _headers: headers,
    };
}

// ── trigger ──────────────────────────────────────────────────────────

test("trigger sets HX-Trigger header with default kind=info", () => {
    const res = fakeRes();
    flash.trigger(res, "Saved.");
    const val = res._headers["HX-Trigger"];
    if (!val) throw new Error("header not set");
    const payload = json.decode(val);
    assertEq(payload.flash.text, "Saved.");
    assertEq(payload.flash.kind, "info");
});

test("trigger honors custom kind", () => {
    const res = fakeRes();
    flash.trigger(res, "Bad input", "error");
    const payload = json.decode(res._headers["HX-Trigger"]);
    assertEq(payload.flash.text, "Bad input");
    assertEq(payload.flash.kind, "error");
});

test("trigger coerces non-string text", () => {
    const res = fakeRes();
    flash.trigger(res, 42);
    const payload = json.decode(res._headers["HX-Trigger"]);
    assertEq(payload.flash.text, "42");
});

test("trigger rejects missing res", () => {
    try { flash.trigger(null, "x"); } catch (_) { return; }
    throw new Error("should have thrown");
});

test("trigger rejects missing text", () => {
    try { flash.trigger(fakeRes(), null); } catch (_) { return; }
    throw new Error("should have thrown");
});

// ── set: input validation (no session needed for the error path) ────

test("set rejects missing text", () => {
    const req = { ctx: { session_id: "fake", session: {} } };
    try {
        flash.set(req, null);
    } catch (e) {
        if (!String(e.message).includes("text is required")) {
            throw new Error(`wrong error: ${e.message}`);
        }
        return;
    }
    throw new Error("should have thrown");
});

test("set errors when no session_id in req.ctx", () => {
    const req = { ctx: {} };
    try {
        flash.set(req, "hi");
    } catch (e) {
        if (!String(e.message).includes("no session")) {
            throw new Error(`wrong error: ${e.message}`);
        }
        return;
    }
    throw new Error("should have thrown");
});

test("set errors when req is null", () => {
    try { flash.set(null, "hi"); } catch (_) { return; }
    throw new Error("should have thrown");
});

// ── consume: safe-defaults ───────────────────────────────────────────

test("consume returns empty array when no session", () => {
    const msgs = flash.consume({ ctx: {} });
    assertEq(Array.isArray(msgs), true);
    assertEq(msgs.length, 0);
});

test("consume returns empty when req is null", () => {
    const msgs = flash.consume(null);
    assertEq(Array.isArray(msgs), true);
    assertEq(msgs.length, 0);
});

export default { pass, fail };
